// node-mailgun/index.js
var http = require('http')
  , https = require('https')
  , _ = require('underscore')
  , noop = function(){};

function Mailer(config) {
  // Force new instance.
  if ( ! (this instanceof Mailer)) {
    return new Mailer(config);
  }
  // Allow string configs.
  if (typeof config !== 'object') { config = { key: config }; }

  // Merge with sensible defaults.
  this.config = { ssl: true };
  _.defaults(this.config, config);
}

// Send the message.
Mailer.prototype.send = function(callback) {
  // Make callback optional.
  if (typeof callback !== 'function') {
    callback = noop;
  }

  // Join multiple recipients.
  var to = _.isArray(this.config.to)
    ? this.config.to.join(',')
    : this.config.to; 
  
  // Prepare post data.
  var post_data = [
    this.config.from
    , to
    , ""
    , "From: "+this.config.from
    , "To: "+to
    , "Content-Type: text/plain;charset=utf-8"
    , "Subject: "+this.config.subject
    , ""
    , this.config.body
  ].join("\n");

  // Prepare request options.
  var options = {
    host: 'mailgun.net'
    , port: this.config.ssl ? 443 : 80
    , path: '/api/messages.eml?servername='+this.config.host
    , method: 'POST'
    , headers: {
      'Authorization': 'Basic '+(new Buffer('api:'+this.config.key).toString("base64"))
      , 'Content-Type': 'text/plain'
      , 'Content-Length': post_data.length
    }
  };

  // Prepare request object.
  (this.config.ssl ? https : http).request(options, function(res) {
    callback(res.statusCode !== 201 ? res.statusCode : null);
  
  // Handle errors.
  }).on('error', function(err) {
    callback(err.message);
  
  // Send request.
  }).end(post_data);
};

// Add message configs.
Mailer.prototype.add = function(key, val) {
  if (typeof key === 'object') {
    _.defaults(this.config, key);
  } else {
    this.config[key] = val;
  }
  return this;
};

// Remove message configs.
Mailer.prototype.remove = function() {
  var args = Array.prototype.slice.call(arguments);
  args.forEach(function(key){
    delete this.config[key];
  });
  return this;
}

module.exports = Mailer;
